include/onnxruntime/core/framework/execution_provider.h

```diff
@@ -96,10 +98,13 @@ class IExecutionProvider {
      contains more than one node. The node indexes contained in sub-graphs may
      have overlap, and it's ONNXRuntime's responsibility to do the partition
      and decide whether a node will be assigned to <*this> execution provider.
+     `kernel_type_str_resolver` can be used to look up kernels in
+     `kernel_registries`.
   */
   virtual std::vector<std::unique_ptr<ComputeCapability>>
   GetCapability(const onnxruntime::GraphViewer& graph_viewer,
-                const std::vector<const KernelRegistry*>& kernel_registries) const;
+                const std::vector<const KernelRegistry*>& kernel_registries,
+                const IKernelTypeStrResolver& kernel_type_str_resolver) const;
```

Comment:
I'm confused. Why do EPs need to know about this resolver stuff?

Reply:
I had a chat with Ed about this. We need to find a way to not pass the resolver in the EP interface. We shouldn't make the EP interface more complicated. Can the resolver be encapsulated in the registry itself? I guess this might not work for ORT format models given we've static registries. Another option is to store the resolver in the KernelRegistryManager (since there's no use case for different kernel registries to have their own resolvers) that can be used by all kernel registries during kernel lookup; needs some thought to work out the details.

Reply:
I'm strongly leaning towards providing a functor so that the EP author knows nothing about multiple registries or resolvers. That leaves us free to change what happens in the background and the EP just looks for a kernel using the Node and the EP name.

Reply:
Providing a functor forces the EP to let the framework decide the order of lookup in that the custom registries will be given preference over the EP static registries. May be this works for most cases. However, shouldn't an EP be given the liberty to decide this? Isn't this why we're calling GetCapability in the first place? If we encapsulate the resolver in each registry it preserves the current semantics while being completely transparent to the EPs.

Reply:
I actually wondering whether we can remove the custom kernel registries from GetCapability API.  My feeling is our design is broken from the beginning. In current design, EP has its own kernel collections, but we allow user to register another kernel registry that contains the custom kernels on the same EP but current EP instance is not aware of it. And later we ask EP to check the capability of those kernels that EP itself even not aware of. It sounds weird. Today although we provide this flexibility, but i believe no EP is using it, they just rely on the default GetCapability check.

IMO, either we keep the design that custom kernels is at session level, so EP instance is not aware of it, and let session decide the capability, or we make those custom kernels registered to the EP instance, so EP is aware of them and knows how to determine to handle it.

Reply:
> Today although we provide this flexibility, but i believe no EP is using it, they just rely on the default GetCapability check.

This is incorrect. We've several users who have written custom cuda kernels; we had to make a change to support them. So, they're indeed getting used. See [this](https://github.com/microsoft/onnxruntime/issues/3138) and [this](https://github.com/microsoft/onnxruntime/pull/3165).

Reply:
right we have custom ops, but those custom ops author actually doesn't expect EP to determine whether those ops are used or not, right? think about if you write a cuda custom op, and you register it to ORT session, do you expect the Cuda EP to determine whether you kernel will be used, or do you expect ORT session to determine whether it will be used?

Reply:
Let's discuss this over email.

Reply:
added an IKernelLookup class to replace the kernel registries and kernel type str resolver parameters, please let me know what you think.

https://github.com/microsoft/onnxruntime/blob/16073fce2b554afa6e4247c24a31b76b93e0c527/include/onnxruntime/core/framework/execution_provider.h#L91-L102